net/socks5/conn.go

```diff
@@ -0,0 +1,151 @@
+package socks5
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"strconv"
+)
+
+type DialContext func(ctx context.Context, network, addr string) (net.Conn, error)
+
+// Conn represents a SOCKS5 connection for client to reach
+// server. The struct is filled by each of the internal methods
+// in turn as the transaction progresses.
+type Conn struct {
+	dialContext DialContext
+	client      net.Conn
+	server      net.Conn
+	methods     []AuthMethod
+	request     *Request
+}
+
+// NewConn creates a new SOCKS5 connection that uses the default
+// dialing context to talk to the SOCKS5 server
+func NewConn(clientConn net.Conn) *Conn {
+	dialer := &net.Dialer{}
+	return NewConnWithDialContext(clientConn, dialer.DialContext)
+}
+
+// NewConnWithDialContext creates a new SOCKS5 connection that uses
+// a custom dialing context to talk to the SOCKS5 server
+func NewConnWithDialContext(clientConn net.Conn, dialContext DialContext) *Conn {
+	return &Conn{
+		client:      clientConn,
+		dialContext: dialContext,
+	}
+}
+
+func (conn *Conn) init() error {
+	buf := make([]byte, MaxInitRequestSize)
+	n, err := conn.client.Read(buf)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Received connection request from %s\n", conn.client.RemoteAddr())
+
+	conn.methods, err = MethodsFromInitPacket(buf[:n])
```

Comment:
MethodsFromInitPacket probably should not be exported.

Reply:
I exported that because I figured it would be useful for someone who might just want a known-good tool to parse a SOCKS5 packet into a struct. If that's not a good idea, let me know and I'll unexport.